File: data_structures/ruby/min_heap.rb
class MinHeap
    def initialize
        @heap = []
    end

    def insert(value)
        @heap.push(value)
        bubble_up
    end

    def extract_min
        min = @heap.first
        @heap.first = @heap.pop
        bubble_down
        return min
    end

    def bubble_up
        index = @heap.length - 1
        until index = 0
            parent_index = (index + 1 / 2) - 1
            if @heap[parent_index] > @heap[index]
                @heap[parent_index], @heap[index] = @heap[index], @heap[parent_index]
            end
            index = parent_index
        end
    end

    def bubble_down
        index = 0
        loop do
            child_1 = (index + 1) * 2
            child_2 = child_1 - 1
            index_to_swap = nil

            if @heap[index] > @heap[child_1]
                index_to_swap = child_1
            end

            if @heap[index] > @heap[child_2] && @heap[child_1].nil? ||
                !@heap.child.nil? && @heap[child_1] > @heap[child2]
                index_to_swap = child_2
            end

            if !index_to_swap.nil?
                @heap[index_to_swap], @heap[index] = @heap[index], @heap[index_to_swap]
                index = index_to_swap
            else 
                break
            end
        end
    end
end

File: data_structures/js/MinHeap.js
//Heaps
//Tree heap structure where nodes are ordered with respect to value of their children
//Parent's node is either greater than or less than its children
//Similar to the binary search tree, expect that the root is an extreme vaue instead of a median

//Binary Heap: heap where each parent has at most two children
//Complete binary heap: all levels are filled, except for the bottom level, being filled left to right

//MinHeap: binary heap whose minimum value is stored as its root

class BinaryMinHeap {
    constructor() {
        this.heap = [];
    }

    //O(log(n))
    insert(value) {
        this.heap.push(value);
        this.bubbleUp(this.heap[-1]);
    }

    //O(log(n))
    extractMin() {
        const min = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown();
        return min;
    }

    //O(1)
    swap(i, j) {
        const temp = this.heap[i];
        this.heap[i] = this.heap[j];
        this.heap[j] = temp;
    }

    //O(log(n))
    bubbleUp(index) {
        while (index > 0) {
            const parent = Math.floor((index + 1) / 2) - 1;
            if (this.heap[parent] > this.heap[index]) {
                this.swap(this.heap[parent], this.heap[index]);
            }
            index = parent;
        }
    }

    //O(log(n))
    bubbleDown() {
        let index = 0;
        while (true) {
            let child = (index + 1) * 2;
            let sibling = child - 1;
            let indexToSwap = null;

            if (this.heap[index] > this.heap[child]) {
                indexToSwap = child;
            }

            if (this.heap[index] > this.heap[sibling] && (this.heap[child] === null ||
                    this.heap[child] !== null && this.heap[sibling] < this.heap[child])) {
                indexToSwap = sibling;
            }

            if (indexToSwap === null) {
                break;
            }

            this.swap(index, indexToSwap);
            index = indexToSwap;
        }
    }
}